partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,275 @@
+    )
+    part.add(clef, 0)
+    for t_start, name, t_end in key_sigs:
+        fifths, mode = key_name_to_fifths_mode(name)
+        t_start, t_end = int(t_start), int(t_end)
+        part.add(score.KeySignature(fifths, mode), t_start, t_end)
+
+    warnings.warn("add notes", stacklevel=2)
+
+    for n in note_array:
+        if n["duration_div"] > 0:
+            note = score.Note(
+                step=n["step"],
+                octave=n["octave"],
+                alter=n["alter"],
+                voice=int(n["voice"] or 0),
+                id=n["id"],
+                symbolic_duration=estimate_symbolic_duration(n["duration_div"], ticks)
+            )
+        else:
+            note = score.GraceNote(
+                grace_type="appoggiatura",
+                step=n["step"],
+                octave=n["octave"],
+                alter=n["alter"],
+                voice=int(n["voice"] or 0),
+                id=n["id"],
+                symbolic_duration=dict(type="quarter"),
+            )
+
+        part.add(note, n["onset_div"], n["onset_div"] + n["duration_div"])
+
+    if np.all(note_array["ts_beats"] == None):
+        warnings.warn("No time signatures found, assuming 4/4")
+        time_sigs = [[0, 4, 4]]
+    else:
+        time_sigs = [[0, note_array[0]["ts_beats"], note_array[0]["ts_beat_type"]]]
+        for n in note_array:
+            if n["ts_beats"] != time_sigs[-1][1] or n["ts_beat_type"] != time_sigs[-1][2]:
+                time_sigs.append([n["onset_div"], n["ts_beats"], n["ts_beat_type"]])
+    time_sigs = np.array(time_sigs)
+
+    # for convenience we add the end times for each time signature
+    ts_end_times = np.r_[time_sigs[1:, 0], np.max(note_array["onset_div"]+note_array["duration_div"])]
+    time_sigs = np.column_stack((time_sigs, ts_end_times))
+
+    warnings.warn("add time sigs and measures", stacklevel=2)
+
+    for ts_start, num, den, ts_end in time_sigs:
+        time_sig = score.TimeSignature(num.item(), den.item())
+        part.add(time_sig, ts_start, ts_end)
+
+    if anacrusis_divs > 0:
+        part.add(score.Measure(0), 0, anacrusis_divs)
+
+    if sanitize:
+        warnings.warn("Inferring measures", stacklevel=2)
+        score.add_measures(part)
+
+        warnings.warn("Find and tie notes", stacklevel=2)
+        # tie notes where necessary (across measure boundaries, and within measures
+        # notes with compound duration)
+        score.tie_notes(part)
+
+        warnings.warn("find and ensure tuplets", stacklevel=2)
+        # apply simplistic tuplet finding heuristic
+        score.find_tuplets(part)
+
+    warnings.warn("done create_part", stacklevel=2)
+    return part
+
+
+def note_array_to_score(note_array: Union[np.ndarray, list], part_id="", divs: int = None, assign_note_ids: bool = True,
+                       ensurelist: bool = False, estimate_key: bool = False, sanitize: bool = True, scorify=True) -> ScoreLike:
```

Comment:
It would be great to have arguments for specifying time and key signatures, in case this information is not provided in the note array, and otherwise, infer this information or set a default. (The most reliable thing would be to estimate the key signature and assign 4/4 as the default time signature).

Reply:
What happens for now, is that if time signature is not available in the note array then it just fails but if we change the clause to something else we can enrich it later. I think it needs maybe a bit more thought on how to do this.